include/circt/AutoGenParser.h

```diff
@@ -0,0 +1,410 @@
+//===- AutoGenParser.h - standard format printer/parser ---------*- C++-*-===//
+//
+// Automatic (template-base) generation of parsers and printers. Defines a
+// standard format for type asm format.
```

Comment:
It is orthogonal to ESI and CIRCT. It was included in my tblgen type definition patch initially, but it was decided that this should be removed from the initial patch and could be reviewed in the future. I figured I'd put it here so we're not so reliant on CIRCT-specific patches to the circt/llvm repo... Should I add there instead?

Either way, it'll eventually go into MLIR in some form. Adding it here means we can move over to the new way in MLIR at our convenience instead of when we update circt/llvm.

Reply:
I guess I still don't understand what this is.  The tablegen ODS printer/parser generator is growing in capability.  Do we need to have a template based approach to do the same thing?  This seems like a case where we should have "one right way to do things".  If your approach is better than ODS, then awesome, lets use it and remove the ODS stuff :-).  We should discuss that on an MLIR forum tho.

Reply:
I forgot to specify: this is just for parsing/printing of types. The new ODS for types (which [just landed today](https://github.com/llvm/llvm-project/commit/5fe53c412832b3c18cd1e4e9f54848304e3a74bc)) doesn't support generation of printer/parsers.  This is a stopgap in the meantime while we (MLIR) figure out how to generate printers/parsers.